### PR TITLE
Add automated scripts

### PR DESCRIPTION
**Pull request by nikosbosse**
add scripts that could be run to update forecasts automatically

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -23,8 +23,9 @@ Imports:
     magrittr,
     data.table,
     dplyr,
-    tibble, 
-    lubridate
+    tibble,
```
Comment (seabbs) [inline DESCRIPTION:26]: scoringutils isn't listed in either Imports or Suggests

File: R/utils.R [hunk 2]
```diff
@@ -78,3 +78,30 @@ globalVariables(
...
+#' Attempt to Execute an Expression and Retry After Failure
+#' #'
+#' @param expr an expression that shell be executed
+#' @param time_to_wait time to wait until the next try after a failure
+#' @param number_of_attempts numeric, how often shall we try?
+#' @return outcome of the expression to be evaluated
+#' @export
+#' @importFrom attempt attempt
+
+try_and_wait <- function(expr,
```
Comment (seabbs) [inline R/utils.R:91]: no imports are listed correctly here and attempt is not in the description.

File: evaluation/forecast-evaluation.R [hunk 3]
```diff
@@ -1,7 +1,6 @@
 # packages ---------------------------------------------------------------------
 library(magrittr)
 
-
 # read in the EpiExpert ensemble forecast and EpiNow2 models
 folders <- list.files("submissions/human-forecasts/")
```
Comment (seabbs) [inline evaluation/forecast-evaluation.R:5]: use here

File: human-forecasts/app.R [hunk 4]
```diff
@@ -18,12 +18,14 @@ first_forecast_date <- submission_date + 5
 # load deaths
 deaths_inc <- data.table::fread("data/weekly-incident-deaths.csv") %>%
```
Comment (seabbs) [inline human-forecasts/app.R:19]: here

File: human-forecasts/redeploy-forecast-app.R [hunk 5]
```diff
@@ -1,5 +1,13 @@
+# update data
+source("data-raw/update.R")
```
Comment (seabbs) [inline human-forecasts/redeploy-forecast-app.R:2]: here

File: human-forecasts/redeploy-forecast-app.R [hunk 6]
```diff
@@ -1,5 +1,13 @@
...
+
+# copy data into app
+file.copy(c("data-raw/weekly-incident-cases.csv", 
+            "data-raw/weekly-incident-deaths.csv"),
+          to = "human-forecasts/data/", overwrite = TRUE)
+
 # reinstall crowdforecastr from github if necessary
-devtools::install_github("epiforecasts/crowdforecastr", force = FALSE)
+devtools::install_github("epiforecasts/crowdforecastr", force = FALSE, upgrade = "always")
```
Comment (seabbs) [inline human-forecasts/redeploy-forecast-app.R:10]: I don't think you need to specify force or upgrade?

Comment (seabbs) [reply human-forecasts/redeploy-forecast-app.R:10]: and shouldn't this be in the DESCRIPTION?

File: human-forecasts/redeploy-forecast-app.R [hunk 7]
```diff
@@ -1,5 +1,13 @@
+# update data
+source("data-raw/update.R")
+
+# copy data into app
+file.copy(c("data-raw/weekly-incident-cases.csv",
```
Comment (seabbs) [inline human-forecasts/redeploy-forecast-app.R:5]: here

File: human-forecasts/redeploy-forecast-app.R [hunk 8]
```diff
@@ -19,8 +27,6 @@ if (weekdays(Sys.Date()) != "Monday") {
 
 saveRDS(submission_date, "human-forecasts/data/submission_date.RDS")
```
Comment (seabbs) [inline human-forecasts/redeploy-forecast-app.R:28]: RDS?

File: human-forecasts/redeploy-forecast-app.R [hunk 9]
```diff
@@ -1,5 +1,13 @@
...
+file.copy(c("data-raw/weekly-incident-cases.csv", 
+            "data-raw/weekly-incident-deaths.csv"),
+          to = "human-forecasts/data/", overwrite = TRUE)
+
 # reinstall crowdforecastr from github if necessary
-devtools::install_github("epiforecasts/crowdforecastr", force = FALSE)
+devtools::install_github("epiforecasts/crowdforecastr", force = FALSE, upgrade = "always")
 
 # helper function to determine the correct forecast date
 next_monday <- function(date){
```
Comment (seabbs) [inline human-forecasts/redeploy-forecast-app.R:13]: change to suggested floor_date approach

File: human-forecasts/redeploy-forecast-app.R [hunk 10]
```diff
@@ -19,8 +27,6 @@ if (weekdays(Sys.Date()) != "Monday") {
 
 saveRDS(submission_date, "human-forecasts/data/submission_date.RDS")
 
-
-
 rsconnect::deployApp(appDir = "human-forecasts/",
```
Comment (seabbs) [inline human-forecasts/redeploy-forecast-app.R:30]: DESCRIPTION

File: human-forecasts/update-submissions.R [hunk 11]
```diff
@@ -0,0 +1,278 @@
+library(googledrive)
```
Comment (seabbs) [inline human-forecasts/update-submissions.R:1]: DESCRIPTION

File: human-forecasts/update-submissions.R [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+drive_auth(cache = ".secrets", email = "epiforecasts@gmail.com")
+gs4_auth(token = drive_token())
+
+spread_sheet <- "1nOy3BfHoIKCHD4dfOtJaz4QMxbuhmEvsWzsrSMx_grI"
+identification_sheet <- "1GJ5BNcN1UfAlZSkYwgr1-AxgsVA2wtwQ9bRwZ64ZXRQ"
+
+
+# setup ------------------------------------------------------------------------
+# - 1 as this is usually updated on a Tuesday
+submission_date <- Sys.Date() - 1
```
Comment (seabbs) [inline human-forecasts/update-submissions.R:18]: Change to Monday detector used elsewhere (feel free to add a utility like `last_monday`)

File: human-forecasts/update-submissions.R [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+  
+  return(forecasts)
+}
+
+# replace time with duration and date with epiweek
+raw_forecasts <- replace_date_and_time(raw_forecasts)
+filtered_forecasts <- replace_date_and_time(filtered_forecasts)
+
+# write raw forecasts
+data.table::fwrite(raw_forecasts %>%
```
Comment (seabbs) [inline human-forecasts/update-submissions.R:87]: not in library list at top of script

File: human-forecasts/update-submissions.R [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+                     dplyr::select(-board_name),
+                   paste0("human-forecasts/raw-forecast-data/", 
+                          submission_date, "-raw-forecasts.csv"))
```
Comment (seabbs) [inline human-forecasts/update-submissions.R:90]: use here

File: human-forecasts/update-submissions.R [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+  tidyr::unnest(cols = c(quantile, value)) %>%
+  dplyr::ungroup() %>%
+  dplyr::mutate(type = ifelse(target_type == "cases", "case", "death"), 
+                target = paste0(horizon, " wk ahead inc ", type), 
+                type = "quantile")
+
+# save forecasts in quantile-format
+data.table::fwrite(forecast_quantiles %>%
+                     dplyr::mutate(submission_date = submission_date),
+                   paste0("human-forecasts/processed-forecast-data/",
```
Comment (seabbs) [inline human-forecasts/update-submissions.R:153]: here

File: human-forecasts/update-submissions.R [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+# add cumulative forecasts -----------------------------------------------------
+
+# get latest cumulative forecast
+first_forecast_date <- forecasts %>%
+  dplyr::pull(target_end_date) %>%
+  as.Date() %>%
+  unique() %>%
+  min(na.rm = TRUE)
+
+deaths <- data.table::fread("data-raw/weekly-cumulative-deaths.csv") %>%
```
Comment (seabbs) [inline human-forecasts/update-submissions.R:213]: here

File: human-forecasts/update-submissions.R [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+  dplyr::inner_join(last_obs) %>%
+  dplyr::mutate(value = value + last_value) %>%
+  dplyr::select(-last_value, -case)
+
+forecast_submission <- dplyr::bind_rows(forecast_inc, forecast_cum) %>%
+  dplyr::mutate(forecast_date = submission_date)
+
+
+# write submission files -------------------------------------------------------
+if (!dir.exists(here::here("submissions", "human-forecasts", submission_date))) {
```
Comment (seabbs) [inline human-forecasts/update-submissions.R:246]: use check_dir

File: rt-crowd-forecast/update-samples.R [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+options(gargle_oauth_cache = ".secrets")
+drive_auth(cache = ".secrets", email = "epiforecasts@gmail.com")
+gs4_auth(token = drive_token())
+
+spread_sheet <- "1g4OBCcDGHn_li01R8xbZ4PFNKQmV-SHSXFlv2Qv79Ks"
+identification_sheet <- "1GJ5BNcN1UfAlZSkYwgr1-AxgsVA2wtwQ9bRwZ64ZXRQ"
+
+# setup ------------------------------------------------------------------------
+# - 1 as this is usually updated on a Tuesday
+submission_date <- Sys.Date() - 1
```
Comment (seabbs) [inline rt-crowd-forecast/update-samples.R:17]: `last_monday`

File: rt-crowd-forecast/update-samples.R [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+                     dplyr::select(-board_name),
+                   paste0("rt-crowd-forecast/raw-forecast-data/", 
+                          submission_date, "-raw-forecasts.csv"))
+
+# draw samples from the distributions ------------------------------------------
+n_people <- filtered_forecasts$forecaster_id %>%
+  unique() %>%
+  length()
+
+draw_samples <- function(distribution,
```
Comment (seabbs) [inline rt-crowd-forecast/update-samples.R:95]: is this the same as in the other script?

File: rt-crowd-forecast/update-samples.R [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+                                            dplyr::rename(prediction = value)) %>%
+  dplyr::mutate(target_end_date = as.Date(target_end_date))
+
+plot <- scoringutils::plot_predictions(check %>%
+                                         dplyr::mutate(true_value = NA_real_, 
+                                                       target_end_date = as.Date(target_end_date,
+                                                                                 origin="1970-01-01")), 
+                                       x = "target_end_date", 
+                                       facet_formula = ~ forecaster_id + location + target_type)
+ggplot2::ggsave(paste0("rt-crowd-forecast/plots/", submission_date, "rt-forecasts.png"))
```
Comment (seabbs) [inline rt-crowd-forecast/update-samples.R:202]: here

**Review by seabbs — COMMENTED** [review]
Some changes to make it function on a remote server without your packages + style.

File: human-forecasts/redeploy-forecast-app.R (hunk 8 above)
Comment (nikosbosse) [reply human-forecasts/redeploy-forecast-app.R:28]: doing that to pass the submission date into the forecast app

Comment (seabbs) [reply human-forecasts/redeploy-forecast-app.R:28]: it should be `.rds`

File: rt-crowd-forecast/update-samples.R (hunk 19 above)
Comment (nikosbosse) [reply rt-crowd-forecast/update-samples.R:95]: similar. in the other script it deals with quantiles, not samples, so not sure it's worth the effort of writing one function for both

**Pull request merged**